// iostreams/iocommon.js

/** IO States Enumeration     */
var KIO = {
  /** IO closed inicial state             */
  closed: 0,
  /** IO connected                        */
  connected: 1,
  /** IO attached to Remote               */
  attached: 2,
  /** IO connecting                       */
  connecting: 3,
  /** IO closing                          */
  closing: 4,
  /** IO close lock                       */
  close_lock: 5,
  /** IO close lock                       */
  close_request: 6,
  /** IO ERROR Server ShutDown            */
  'server_shutdown': 200,
  /** IO ERROR Server Timeout             */
  'server_timeout': 201,
  /** IO ERROR Socket Timeout             */
  'socket_timeout': 202,
  /** IO ERROR Previous Name exists       */
  'name_exist': 203,
  /** IO ERROR Closed by request          */
  'closed_by_request': 204,
  /** IO ERROR Bad Protocol               */
  'bad_protocol': 205,
  /** IO ERROR Socket is closed           */
  'socket_closed': 206,
  /** Map Key to denote bridged Address   */
  bridgedIPKey: '@ip',
  /** Map Key to denote bridged Data      */
  bridgedDataKey: '@data',
  /** Maximum Keep Alive time in seconds  */
  maxKeepAlive: 300,
  /** Default Close reason                */
  defaultClose: {
    'action' : 'disconnect',
    'reason': this.socket_closed,
    'message': 'SocketClosed'
  },
  /** Time in milliseconds to 
   * full dispose a stream               */
  closeDelay: 1000
};
/**
 * Default Close Message
 */
const DefaultCloseReason = {
  'action' : 'disconnect',
  'reason': KIO.socket_closed,
  'message': 'SocketClosed'
};

module.exports = {
  DefaultCloseReason,
  KIO
};
